pkg/binding/bindingpolicy-resolution-broker.go

```diff
@@ -0,0 +1,129 @@
+	// Destinations is a list of destinations that are the "where" part of the
+	// resolution.
+	Destinations []v1alpha1.Destination
+	// ObjectIdentifierToData is a map of object identifiers to their
+	// corresponding ObjectData.
+	ObjectIdentifierToData map[util.ObjectIdentifier]struct {
+		ResourceVersion  string
+		StatusCollectors []string
+	}
+}
+
+// NewResolutionBroker creates a new ResolutionBroker with the given
+// bindingPolicyResolutionGetter function.
+// The latter function's returned `bindingPolicyResolution` is expected to be
+// thread-safe under the constraint that only its methods are used.
+func NewResolutionBroker(bindingPolicyResolutionGetter func(bindingPolicyKey string) *bindingPolicyResolution) ResolutionBroker {
+	return &resolutionBroker{
+		bindingPolicyResolutionGetter: bindingPolicyResolutionGetter,
+	}
+}
+
+// resolutionBroker implements the ResolutionBroker interface.
+// The broker requires a bindingPolicyResolutionGetter function that retrieves
+// resolutions from a BindingPolicyResolver.
+type resolutionBroker struct {
+	sync.Mutex
+	callbacks []func(bindingPolicyKey string)
+	// bindingPolicyResolutionGetter is a function that retrieves the
+	// resolution for a given bindingPolicyKey. If no resolution is associated
+	// with the given key, nil is returned.
+	// The returned `bindingPolicyResolution` is expected to be thread-safe as
+	// long as only its methods are used.
+	bindingPolicyResolutionGetter func(bindingPolicyKey string) *bindingPolicyResolution
+}
+
+// RegisterCallback adds a new callback function that will be called, on a
+// separate goroutine, when a resolution is updated. The callback function
+// should accept a bindingPolicyKey.
+// There is no deduplication of callbacks.
+func (broker *resolutionBroker) RegisterCallback(callback func(bindingPolicyKey string)) {
+	broker.Lock()
+	defer broker.Unlock()
+
+	broker.callbacks = append(broker.callbacks, callback)
+}
+
+// GetResolution retrieves the resolution for a given bindingPolicyKey.
+// If no resolution is associated with the given key, nil is returned.
+func (broker *resolutionBroker) GetResolution(bindingPolicyKey string) *Resolution {
+	bindingPolicyResolution := broker.bindingPolicyResolutionGetter(bindingPolicyKey)
+
+	if bindingPolicyResolution == nil {
+		return nil
+	}
+
+	return &Resolution{
+		Destinations: bindingPolicyResolution.getDestinationsList(),
+		ObjectIdentifierToData: abstract.PrimitiveMapSafeMap(&bindingPolicyResolution.RWMutex,
+			bindingPolicyResolution.objectIdentifierToData,
+			func(data *objectData) struct {
+				ResourceVersion  string
+				StatusCollectors []string
+			} {
+				return struct {
+					ResourceVersion  string
+					StatusCollectors []string
+				}{
+					ResourceVersion:  data.ResourceVersion,
+					StatusCollectors: sets.List(data.StatusCollectors), // members are string copies
+				}
+			}), // while this function breaks the constraint, it maintains its own concurrency safety
```

Comment:
FYI, GetResolution would be more succinct if this func were be a method of `*objectData`.